// renovate/dist/modules/datasource/docker/dockerhub-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DockerHubCache = void 0;
const tslib_1 = require("tslib");
const dequal_1 = require("dequal");
const luxon_1 = require("luxon");
const packageCache = tslib_1.__importStar(require("../../../util/cache/package"));
const cacheNamespace = 'datasource-docker-hub-cache';
class DockerHubCache {
    dockerRepository;
    cache;
    isChanged = false;
    constructor(dockerRepository, cache) {
        this.dockerRepository = dockerRepository;
        this.cache = cache;
    }
    static async init(dockerRepository) {
        let repoCache = await packageCache.get(cacheNamespace, dockerRepository);
        repoCache ??= {
            items: {},
            updatedAt: null,
        };
        return new DockerHubCache(dockerRepository, repoCache);
    }
    reconcile(items) {
        let needNextPage = true;
        let { updatedAt } = this.cache;
        let latestDate = updatedAt ? luxon_1.DateTime.fromISO(updatedAt) : null;
        for (const newItem of items) {
            const id = newItem.id;
            const oldItem = this.cache.items[id];
            if ((0, dequal_1.dequal)(oldItem, newItem)) {
                needNextPage = false;
                continue;
            }
            this.cache.items[newItem.id] = newItem;
            const newItemDate = luxon_1.DateTime.fromISO(newItem.last_updated);
            if (!latestDate || latestDate < newItemDate) {
                updatedAt = newItem.last_updated;
                latestDate = newItemDate;
            }
            this.isChanged = true;
        }
        this.cache.updatedAt = updatedAt;
        return needNextPage;
    }
    async save() {
        if (this.isChanged) {
            await packageCache.set(cacheNamespace, this.dockerRepository, this.cache, 3 * 60 * 24 * 30);
        }
    }
    getItems() {
        return Object.values(this.cache.items);
    }
}
exports.DockerHubCache = DockerHubCache;
