// renovate/dist/modules/manager/pub/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.categories = exports.defaultConfig = exports.supportsLockFileMaintenance = exports.supportedDatasources = exports.extractPackageFile = exports.updateArtifacts = void 0;
const tslib_1 = require("tslib");
const dart_1 = require("../../datasource/dart");
const dart_version_1 = require("../../datasource/dart-version");
const flutter_version_1 = require("../../datasource/flutter-version");
const npmVersioning = tslib_1.__importStar(require("../../versioning/npm"));
var artifacts_1 = require("./artifacts");
Object.defineProperty(exports, "updateArtifacts", { enumerable: true, get: function () { return artifacts_1.updateArtifacts; } });
var extract_1 = require("./extract");
Object.defineProperty(exports, "extractPackageFile", { enumerable: true, get: function () { return extract_1.extractPackageFile; } });
exports.supportedDatasources = [
    dart_1.DartDatasource.id,
    dart_version_1.DartVersionDatasource.id,
    flutter_version_1.FlutterVersionDatasource.id,
];
exports.supportsLockFileMaintenance = true;
exports.defaultConfig = {
    fileMatch: ['(^|/)pubspec\\.ya?ml$'],
    versioning: npmVersioning.id,
};
exports.categories = ['dart'];
