// renovate/dist/logger/remap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetRepositoryLogLevelRemaps = exports.setRepositoryLogLevelRemaps = exports.resetGlobalLogLevelRemaps = exports.setGlobalLogLevelRemaps = exports.getRemappedLevel = void 0;
const string_match_1 = require("../util/string-match");
let globalRemaps;
let repositoryRemaps;
let matcherCache = new WeakMap();
function match(remap, input) {
    const { matchMessage: pattern } = remap;
    let matchFn = matcherCache.get(remap);
    if (!matchFn) {
        matchFn = (0, string_match_1.getRegexOrGlobPredicate)(pattern);
        matcherCache.set(remap, matchFn);
    }
    return matchFn(input);
}
function getRemappedLevel(msg) {
    if (repositoryRemaps) {
        for (const remap of repositoryRemaps) {
            if (match(remap, msg)) {
                return remap.newLogLevel;
            }
        }
    }
    if (globalRemaps) {
        for (const remap of globalRemaps) {
            if (match(remap, msg)) {
                return remap.newLogLevel;
            }
        }
    }
    return null;
}
exports.getRemappedLevel = getRemappedLevel;
function resetMatcherCache() {
    matcherCache = new WeakMap();
}
function setGlobalLogLevelRemaps(remaps) {
    globalRemaps = remaps;
}
exports.setGlobalLogLevelRemaps = setGlobalLogLevelRemaps;
function resetGlobalLogLevelRemaps() {
    globalRemaps = undefined;
    resetMatcherCache();
}
exports.resetGlobalLogLevelRemaps = resetGlobalLogLevelRemaps;
function setRepositoryLogLevelRemaps(remaps) {
    repositoryRemaps = remaps;
}
exports.setRepositoryLogLevelRemaps = setRepositoryLogLevelRemaps;
function resetRepositoryLogLevelRemaps() {
    repositoryRemaps = undefined;
    resetMatcherCache();
}
exports.resetRepositoryLogLevelRemaps = resetRepositoryLogLevelRemaps;
